### ISSUE TITLE
akka-streams, downstream failure accross an async boundary may be communicated as a cancellation 

### ISSUE THREAD

**Issue by eyalfa** [akka/akka#29194]
in graphs including async boundaries, when downstream fails the errors are supposed to be propagated upstream (by onDownstreamFinish(cause)`).
however, when downstream fails before subscription to upstream is complete, upstream is notified with a non failure cancellation,

the following two tests demonstrate this (the one with the `zip` closely resembles an issue we've seen in production), both use `Thread.sleep` to block the `preStart` call which effectively prevents the interpreter actor from processing subscriptions.

```
test("akka"){
    val slowInitSrc = Source
      .unfoldResource( () => {Thread.sleep(5000)},
        (u : Unit) => Some(1),
        (u : Unit) => ()
      )
      .watchTermination()(Keep.right)
      .async

    val failingSrc = Source
      .failed(new RuntimeException("I hate mondays"))
      .watchTermination()(Keep.right)

    val g = slowInitSrc
      .zipMat(failingSrc)(Keep.both)
      .to(Sink.ignore)

    val (f1, f2) = g.run()
    f2.failed.futureValue should have message("I hate mondays")
    Thread.sleep(5000)
    f1.failed.futureValue should have message("I hate mondays")
  }
```

```
test("akka2"){
    val slowInitSrc = Source
      .unfoldResource( () => {Thread.sleep(5000)},
        (u : Unit) => Some(1),
        (u : Unit) => ()
      )
      .watchTermination()(Keep.right)
      .async
    val downstream = Flow[Int]
      .prepend(Source single 1)
      .flatMapPrefix(0){
        case Nil => sys error "I hate mondays"
      }
      .watchTermination()(Keep.right)
      .to(Sink.ignore)

    val g = slowInitSrc
      .toMat(downstream)(Keep.both)

    val (f1, f2) = g.run()
    f2.failed.futureValue should have message("I hate mondays")
    Thread.sleep(5000)
    f1.failed.futureValue should have message("I hate mondays")
  }
```

I suspect the root cause for this is in `akka.stream.impl.fusing.ActorGraphInterpreter.BatchingActorInputBoundary#onDownstreamFinish` which delegates to `akka.stream.impl.fusing.ActorGraphInterpreter.BatchingActorInputBoundary#cancel` which sets `upstreamCompleted = true` which 'confuses' `akka.stream.impl.fusing.ActorGraphInterpreter.BatchingActorInputBoundary#onSubscribe` when the subscription finally arrives.

I'd be happy to open a pull-request for this if it's indeed a bug.

**Comment by jrudolph** [comment]
Thanks for the report.

> `akka.stream.impl.fusing.ActorGraphInterpreter.BatchingActorInputBoundary#cancel` which sets `upstreamCompleted = true` which 'confuses'

Yes, indeed. Maybe it's enough to move that `upstreamCompleted = true` into the `if (upstream != null)` block?

**Comment by jrudolph** [comment]
> I'd be happy to open a pull-request for this if it's indeed a bug.

Thanks, that would be nice. Probably some tests will fail, I didn't review all of the test suite when that behavior was changed, so I probably missed that problem.

**Comment by eyalfa** [comment]
sure, I'll start sketching the pr later today

**Comment by eyalfa** [comment]
@jrudolph , can u point me to the right suite for these two tests?

**Comment by jrudolph** [comment]
> @jrudolph , can u point me to the right suite for these two tests?
> and, if there's a specific test suite for the boundary stages as well.

Actually, no. Let's try if something fails if that change is made and fix those tests if necessary. If nothing (at least) we should add new tests.

**Comment by eyalfa** [comment]
> > @jrudolph , can u point me to the right suite for these two tests?
> > and, if there's a specific test suite for the boundary stages as well.
> 
> Actually, no. Let's try if something fails if that change is made and fix those tests if necessary. If nothing (at least) we should add new tests.

these two tests explicitly test the issue, I think introducing  them  as failing test, then fixing, then looking for  regressions is  a better approach here,, WDUT?

**Comment by jrudolph** [comment]
> these two tests explicitly test the issue, I think introducing them as failing test, then fixing, then looking for regressions is a better approach here,, WDUT?

Yes, makes sense. I had another look and `FusingSpec` seems to contain the tests that check the behavior of `async` explicitly, so that might be the right place to put them into.

**Comment by jrudolph** [comment]
Fixed in #29213. Thanks a lot, @eyalfa!
